### PR TITLE
support free gas for new bridge account

### PR DESCRIPTION
**Pull request by chengzhinei**

### CODE REVIEW

File: zk/txpool/pool_xlayer.go [hunk 1]
```diff
@@ -45,19 +53,33 @@ func (p *TxPool) checkWhiteAddr(addr common.Address) bool {
...
-
-func (p *TxPool) SetGpCacheForXLayer(gpCache GPCache) {
-	p.gpCache = gpCache
+func (p *TxPool) checkFreeGasAddr(senderID uint64) (bool, bool) {
+	addr, ok := p.senders.senderID2Addr[senderID]
+	if !ok {
+		return false, false
+	}
+	// is claim tx
+	for _, e := range p.wbCfg.FreeClaimGasAddr {
```
Comment (sieniven) [inline zk/txpool/pool_xlayer.go:78]: We should also change FreeClaimGasAddr to plural
```suggestion
	for _, e := range p.wbCfg.FreeClaimGasAddrs {
```

File: zk/txpool/pool.go [hunk 2]
```diff
@@ -381,7 +382,12 @@ func New(newTxs chan types.Announcements, coreDB kv.RoDB, cfg txpoolcfg.Config,
 			BlockedList:      ethCfg.DeprecatedTxPool.BlockedList,
 			FreeClaimGasAddr: ethCfg.DeprecatedTxPool.FreeClaimGasAddr,
 			GasPriceMultiple: ethCfg.DeprecatedTxPool.GasPriceMultiple,
+			EnableFreeGasByNonce: ethCfg.DeprecatedTxPool.EnableFreeGasByNonce,
+			FreeGasExAddress:     ethCfg.DeprecatedTxPool.FreeGasExAddress,
+			FreeGasCountPerAddr:  ethCfg.DeprecatedTxPool.FreeGasCountPerAddr,
+			FreeGasLimit:         ethCfg.DeprecatedTxPool.FreeGasLimit,
 		},
+		freeGasAddress: map[string]bool{},
```
Comment (sieniven) [inline zk/txpool/pool.go:390]: Change to plural
```suggestion
		freeGasAddrs: map[string]bool{},
```

File: zk/txpool/pool.go [hunk 3]
```diff
@@ -381,7 +382,12 @@ func New(newTxs chan types.Announcements, coreDB kv.RoDB, cfg txpoolcfg.Config,
 			BlockedList:      ethCfg.DeprecatedTxPool.BlockedList,
 			FreeClaimGasAddr: ethCfg.DeprecatedTxPool.FreeClaimGasAddr,
 			GasPriceMultiple: ethCfg.DeprecatedTxPool.GasPriceMultiple,
+			EnableFreeGasByNonce: ethCfg.DeprecatedTxPool.EnableFreeGasByNonce,
+			FreeGasExAddress:     ethCfg.DeprecatedTxPool.FreeGasExAddress,
+			FreeGasCountPerAddr:  ethCfg.DeprecatedTxPool.FreeGasCountPerAddr,
+			FreeGasLimit:         ethCfg.DeprecatedTxPool.FreeGasLimit,
```
Comment (sieniven) [inline zk/txpool/pool.go:388]: Fmt indentation

File: zk/txpool/pool_xlayer.go [hunk 4]
```diff
@@ -45,19 +53,33 @@ func (p *TxPool) checkWhiteAddr(addr common.Address) bool {
 	return false
 }
 
-func (p *TxPool) isFreeClaimAddr(senderID uint64) bool {
+func (p *TxPool) SetGpCacheForXLayer(gpCache GPCache) {
+	p.gpCache = gpCache
+}
+
+func (p *TxPool) checkFreeGasExAddress(senderID uint64) bool {
```
Comment (sieniven) [inline zk/txpool/pool_xlayer.go:60]: Keep your function naming consistent
```suggestion
func (p *TxPool) checkFreeGasExAddr(senderID uint64) bool {
```

File: zk/txpool/pool_zk.go [hunk 5]
```diff
@@ -62,6 +67,25 @@ func (p *TxPool) onSenderStateChange(senderID uint64, senderNonce uint64, sender
 			toDel = append(toDel, mt)
 			return true
 		}
+		// parse claim tx or dex tx, and add the withdraw addr into free gas cache
+		if p.checkFreeGasExAddress(senderID) {
+			inputHex := hex.EncodeToHex(mt.Tx.Rlp)
+			if strings.HasPrefix(inputHex, "0xa9059cbb") && len(inputHex) > 74 {
+				addrHex := "0x" + inputHex[10:74]
```
Comment (sieniven) [inline zk/txpool/pool_zk.go:74]: Hardcoding again here. Not sure but why cant we use the senderID2Addr map to get the sender's address here to be added into the freeGasAddr map?

File: zk/txpool/pool_zk.go [hunk 6]
```diff
@@ -62,6 +67,25 @@ func (p *TxPool) onSenderStateChange(senderID uint64, senderNonce uint64, sender
...
+				p.freeGasAddress[addrHex] = true
+			} else {
+				txnDec, err := core_types.DecodeTransaction(rlp.NewStream(bytes.NewReader(mt.Tx.Rlp), uint64(len(mt.Tx.Rlp))))
```
Comment (sieniven) [inline zk/txpool/pool_zk.go:77]: Why is there a need to decode the raw transaction again? Doesnt metaTx contain all the necessary information for the pool?

File: zk/txpool/pool_zk.go [hunk 7]
```diff
@@ -62,6 +67,25 @@ func (p *TxPool) onSenderStateChange(senderID uint64, senderNonce uint64, sender
 			toDel = append(toDel, mt)
 			return true
 		}
+		// parse claim tx or dex tx, and add the withdraw addr into free gas cache
+		if p.checkFreeGasExAddress(senderID) {
+			inputHex := hex.EncodeToHex(mt.Tx.Rlp)
+			if strings.HasPrefix(inputHex, "0xa9059cbb") && len(inputHex) > 74 {
```
[the remark this answers is not in the document]
Comment (chengzhinei) [reply zk/txpool/pool_zk.go:75]: This is dex transaction. The from is dex addr, call data is a contract ( from -> to ). So the contract is constant.

File: zk/txpool/pool_zk.go (hunk 5 above)
Comment (chengzhinei) [reply zk/txpool/pool_zk.go:76]: This is dex transaction. The from is dex addr, call data is a contract ( from -> to ). The from addr only can get from tx.data

File: zk/txpool/pool_zk.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (chengzhinei) [reply zk/txpool/pool_zk.go:73]: It's same code with node

File: zk/txpool/pool_zk.go [hunk 8]
```diff
@@ -62,6 +67,25 @@ func (p *TxPool) onSenderStateChange(senderID uint64, senderNonce uint64, sender
...
+				txnDec, err := core_types.DecodeTransaction(rlp.NewStream(bytes.NewReader(mt.Tx.Rlp), uint64(len(mt.Tx.Rlp))))
+				to := txnDec.GetTo()
+				if err == nil && to != nil {
+					p.freeGasAddress[to.Hex()] = true
+				}
+
+			}
+		} else if claim && mt.Tx.Nonce < p.wbCfg.FreeGasCountPerAddr {
+			inputHex := hex.EncodeToHex(mt.Tx.Rlp)
+			addrHex := "0x" + inputHex[4490:4554]
```
[the remark this answers is not in the document]
Comment (chengzhinei) [reply zk/txpool/pool_zk.go:81]: It's a contract as well, the tx is SeqAddr -> bridgeAddr, the call data is ( claim to UserAddr ), we should get the UserAddr, only can get by tx.data.
Same code with node.

**Comment by zjg555543** [comment]
Please add some e2e tests for this PR

**Review by zjg555543 — APPROVED** [review]

File: ethclient/ethclient.go [hunk 9]
```diff
@@ -41,6 +42,29 @@ type Client struct {
...
+	var br *polygonzkevmbridge.Polygonzkevmbridge
+	if len(bridgeSCAddr) != 0 {
+		br, err = polygonzkevmbridge.NewPolygonzkevmbridge(bridgeSCAddr, client)
+	}
+	return &L1Client{
+		Client: client,
+		Bridge: br,
+	}, err
+}
+
```
Comment (sieniven) [inline ethclient/ethclient.go:67]: Is this L1Client only going to be used on e2e test?

Comment (sieniven) [reply ethclient/ethclient.go:67]: Can we please remove this entire code here? Dont think it is necessary to maintain it here if it is used purely for e2e testing. We can instantiate the bridge SC instance inside sendBridgeAsset, then send the contract call tx within that method itself

Comment (chengzhinei) [reply ethclient/ethclient.go:67]: Yes by now. If there is no else use to bridge.
So maybe move it to the e2e package?

Comment (chengzhinei) [reply ethclient/ethclient.go:67]: > Is this L1Client only going to be used on e2e test?

Yes by now. If there is no else use to bridge.
So maybe move it to the e2e package?

File: test/e2e/smoke_test.go [hunk 10]
```diff
@@ -367,3 +461,22 @@ func TestMinGasPrice(t *testing.T) {
...
+		destAddr = &auth.From
+	}
+	tx, err := c.Bridge.BridgeAsset(auth, destNetwork, *destAddr, amount, tokenAddr, true, metadata)
+	if err != nil {
+		return err
+	}
+	// wait transfer to be included in a batch
+	const txTimeout = 60 * time.Second
+	return operations.WaitTxToBeMined(ctx, c.Client, tx, txTimeout)
+}
```
Comment (sieniven) [inline test/e2e/smoke_test.go:482]:
```suggestion
func sendBridgeAsset(ctx context.Context, tokenAddr common.Address, amount *big.Int, destNetwork uint32,
	destAddr *common.Address, metadata []byte, auth *bind.TransactOpts, bridgeSCAddr common.Address, c *ethclient.Client,
) error {
	emptyAddr := common.Address{}
	if tokenAddr == emptyAddr {
		auth.Value = amount
	}
	if destAddr == nil {
		destAddr = &auth.From
	}
	if len(bridgeSCAddr) != 0 {
		return fmt.Errorf("Bridge address error")
	}

	br, err := polygonzkevmbridge.NewPolygonzkevmbridge(bridgeSCAddr, c)
	if err != nil {
		return err
	}
	tx, err := br.BridgeAsset(auth, destNetwork, *destAddr, amount, tokenAddr, true, metadata)
	if err != nil {
		return err
	}

	// wait transfer to be included in a batch
	const txTimeout = 60 * time.Second
	return operations.WaitTxToBeMined(ctx, c.Client, tx, txTimeout)
}
```

File: ethclient/ethclient.go (hunk 9 above)
Comment (zjg555543) [reply ethclient/ethclient.go:67]: L1Client -> BridgeClient

File: test/e2e/smoke_test.go (hunk 10 above)
Comment (zjg555543) [reply test/e2e/smoke_test.go:482]: I think it necessary to modify this code.

**Review by giskook — APPROVED** [review]

File: cmd/utils/flags_xlayer.go [hunk 11]
```diff
@@ -274,6 +291,23 @@ func setTxPoolXLayer(ctx *cli.Context, cfg *ethconfig.DeprecatedTxPoolConfig) {
 	if ctx.IsSet(TxPoolGasPriceMultiple.Name) {
 		cfg.GasPriceMultiple = ctx.Uint64(TxPoolGasPriceMultiple.Name)
 	}
+	if ctx.IsSet(TxPoolEnableFreeGasByNonce.Name) {
+		cfg.EnableFreeGasByNonce = ctx.Bool(TxPoolEnableFreeGasByNonce.Name)
+	}
+	if ctx.IsSet(TxPoolFreeGasExAddrs.Name) {
+		addrHexes := SplitAndTrim(ctx.String((TxPoolFreeGasExAddrs.Name)))
```
Comment (sieniven) [inline cmd/utils/flags_xlayer.go:298]:
```suggestion
		addrHexes := SplitAndTrim(ctx.String(TxPoolFreeGasExAddrs.Name))
```

File: test/e2e/smoke_test.go [hunk 12]
```diff
@@ -25,6 +28,29 @@ const (
...
+	var br *polygonzkevmbridge.Polygonzkevmbridge
+	if len(bridgeSCAddr) != 0 {
+		br, err = polygonzkevmbridge.NewPolygonzkevmbridge(bridgeSCAddr, client)
+	}
+	return &L1Client{
+		Client: client,
+		Bridge: br,
+	}, err
+}
+
```
Comment (sieniven) [inline test/e2e/smoke_test.go:53]: I thought the whole idea was to add this into sendBridgeAsset instead?

File: test/e2e/smoke_test.go [hunk 13]
```diff
@@ -367,3 +485,22 @@ func TestMinGasPrice(t *testing.T) {
...
+		destAddr = &auth.From
+	}
+	tx, err := c.Bridge.BridgeAsset(auth, destNetwork, *destAddr, amount, tokenAddr, true, metadata)
+	if err != nil {
+		return err
+	}
+	// wait transfer to be included in a batch
+	const txTimeout = 60 * time.Second
+	return operations.WaitTxToBeMined(ctx, c.Client, tx, txTimeout)
+}
```
Comment (sieniven) [inline test/e2e/smoke_test.go:506]: https://github.com/okx/xlayer-erigon/pull/69#discussion_r1718006249

File: zk/txpool/pool_xlayer.go [hunk 14]
```diff
@@ -43,18 +51,39 @@ func (p *TxPool) SetApolloConfig(cfg ApolloConfig) {
...
 	if !ok {
 		return false
 	}
-	return p.apolloCfg.CheckFreeClaimAddr(p.xlayerCfg.FreeClaimGasAddrs, addr)
+	for _, e := range p.xlayerCfg.FreeGasExAddrs {
+		if common.HexToAddress(e) == addr {
+			return true
+		}
+	}
+	return false
```
Comment (sieniven) [inline zk/txpool/pool_xlayer.go:68]: Merge conflict with new upstream changes here

Comment (sieniven) [reply zk/txpool/pool_xlayer.go]: Resolved in [219830d](https://github.com/okx/xlayer-erigon/pull/69/commits/219830d646bdaea22e9191b22696b8098bcf43f1)

File: test/e2e/smoke_test.go (hunk 13 above)
Comment (sieniven) [reply test/e2e/smoke_test.go]: Resolved in [219830d](https://github.com/okx/xlayer-erigon/pull/69/commits/219830d646bdaea22e9191b22696b8098bcf43f1)

**Review by sieniven — APPROVED** [review]

**Review by sieniven — APPROVED** [review]

**Pull request merged**